### PR TITLE
Enable dynamic scheduling

### PR DESCRIPTION
**Pull request by Andyli1007**
- Refactoring to unify the code to support both nccl and rccl platform. 
- Support dynamic scheduling based on the platform detected.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,6 +1,46 @@
...
+$ apt install build/pkg/deb/libmsccl-scheduler1_1.0.0-1+cuda._amd64.deb
+```
+
+## Usage
+
+When running applications using MSCCL, set the following environmental variables accordingly:
+1. Add path to the built binary of this scheduler to `LD_PRELOAD`.
+2. Set environment variable 
+   **- Nccl:** `NCCL_MSCCL_ENABLE` to 1.   
+   **- Rccl**: Set `RCCL_MSCCL_ENABLE` to 1.
```
Comment (yzygitzh) [inline README.md:42]: Should be aligned with line 41

File: src/scheduler.cc [hunk 2]
```diff
@@ -9,23 +9,71 @@
...
 
 static const char* mscclAlgoDirEnv = "MSCCL_ALGO_DIR";
 static const char* mscclAlgoDefaultDir = "msccl-algorithms";
 extern "C" bool mscclUnitTestMode() __attribute__((__weak__));
 static const char* mscclUnitTestAlgoDefaultDir = "msccl-unit-test-algorithms";
-static const char* mscclAlgoShareDirPath = "share/rccl/msccl-algorithms";
-static const char* mscclUnitTestAlgoShareDirPath = "share/rccl/msccl-unit-test-algorithms";
+static const char* mscclAlgoShareDirPath = "share/msccl-scheduler/msccl-algorithms";
+static const char* mscclUnitTestAlgoShareDirPath = "..share/msccl-unit-test-algorithms";
+static const char* mscclAzureVMDetectionAgent = "http://169.254.169.254/metadata/instance?api-version=2019-06-04";
```
Comment (yzygitzh) [inline src/scheduler.cc:33]: It cannot auto-detect A100 40G NDv4. Shall we add a warning message for user when auto-detection fails to find a matching VM?

File: src/scheduler.cc [hunk 3]
```diff
@@ -9,23 +9,71 @@
...
+#define MSCCL_SCHEDULER_NAME "github.com/Azure/msccl-scheduler"
 
 static const char* mscclAlgoDirEnv = "MSCCL_ALGO_DIR";
 static const char* mscclAlgoDefaultDir = "msccl-algorithms";
 extern "C" bool mscclUnitTestMode() __attribute__((__weak__));
 static const char* mscclUnitTestAlgoDefaultDir = "msccl-unit-test-algorithms";
-static const char* mscclAlgoShareDirPath = "share/rccl/msccl-algorithms";
-static const char* mscclUnitTestAlgoShareDirPath = "share/rccl/msccl-unit-test-algorithms";
+static const char* mscclAlgoShareDirPath = "share/msccl-scheduler/msccl-algorithms";
+static const char* mscclUnitTestAlgoShareDirPath = "..share/msccl-unit-test-algorithms";
```
Comment (yzygitzh) [inline src/scheduler.cc:32]: Should be ../share?

File: src/scheduler.cc [hunk 4]
```diff
@@ -35,27 +83,28 @@ __hidden ncclResult_t mscclSchedulerInit() {
...
-    // Try to find default algorithm directory based on librccl.so path
+    // Try to find default algorithm directory based on scheduler.so and shcheduler algo installtion path.
     Dl_info dl_info;
     struct link_map *link_map_ptr = nullptr;
-    if (!dladdr1((void *)ncclAllReduce, &dl_info, (void **)&link_map_ptr, RTLD_DL_LINKMAP)) {
+    if (!dladdr1((void *)mscclSchedulerInit, &dl_info, (void **)&link_map_ptr, RTLD_DL_LINKMAP)) {
       fprintf(stderr, "%s: dladdr1 failed", MSCCL_SCHEDULER_NAME);
       return ncclInvalidUsage;
     }
     std::string selfLibPath = link_map_ptr->l_name;
```
Comment (yzygitzh) [inline src/scheduler.cc:93]: Shall we keep aligned with https://github.com/ROCmSoftwarePlatform/rccl/blob/develop/src/misc/msccl/msccl_lifecycle.cc#L96-L103 , especially we should replace the second "find" to "find_last_of". This is a known bug that has been fixed in rccl.

File: src/scheduler.cc [hunk 5]
```diff
@@ -9,23 +9,71 @@
...
-#define MSCCL_SCHEDULER_NAME "github.com/microsoft/msccl-scheduler"
+#define MSCCL_SCHEDULER_NAME "github.com/Azure/msccl-scheduler"
 
 static const char* mscclAlgoDirEnv = "MSCCL_ALGO_DIR";
 static const char* mscclAlgoDefaultDir = "msccl-algorithms";
 extern "C" bool mscclUnitTestMode() __attribute__((__weak__));
 static const char* mscclUnitTestAlgoDefaultDir = "msccl-unit-test-algorithms";
-static const char* mscclAlgoShareDirPath = "share/rccl/msccl-algorithms";
-static const char* mscclUnitTestAlgoShareDirPath = "share/rccl/msccl-unit-test-algorithms";
+static const char* mscclAlgoShareDirPath = "share/msccl-scheduler/msccl-algorithms";
```
Comment (yzygitzh) [inline src/scheduler.cc:31]: Should be ../share?

Comment (yzygitzh) [reply src/scheduler.cc:31]: Ref: https://github.com/ROCmSoftwarePlatform/rccl/blob/develop/src/misc/msccl/msccl_lifecycle.cc#L78-L79

File: src/scheduler.cc [hunk 6]
```diff
@@ -9,23 +9,71 @@
...
+        curl_easy_setopt(curl, CURLOPT_URL, mscclAzureVMDetectionAgent);
+        struct curl_slist *headers = NULL;
+        headers = curl_slist_append(headers, "Metadata:true");
+        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
+        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, writeCallback);
+        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &readBuffer);
+        curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT, 1L);
+        res = curl_easy_perform(curl);
+        if(res != CURLE_OK) {
+            fprintf(stderr, "MSCCL Get Azure VM Size failed: %s", curl_easy_strerror(res));
```
Comment (yzygitzh) [inline src/scheduler.cc:62]: Logs from MSCCL scheduler should use its own tag, line 90 is a good example (fprintf(stderr, "%s: dladdr1 failed", MSCCL_SCHEDULER_NAME);)

File: src/scheduler.cc [hunk 7]
```diff
@@ -35,27 +83,28 @@ __hidden ncclResult_t mscclSchedulerInit() {
...
+    mscclAlgoShareDirStr += (mscclUnitTestMode && mscclUnitTestMode()) ? mscclUnitTestAlgoShareDirPath : updateAlgoDirByVMSize(std::string(mscclAlgoShareDirPath));
     mscclAlgoShareDir = mscclAlgoShareDirStr.c_str();
   }
+  fprintf(stdout, "MSCCL: External Scheduler will use %s as algorithm directory and %s as share algorithm directory\n", mscclAlgoDir, mscclAlgoShareDir);
   struct dirent *entry = nullptr;
   DIR *dp = nullptr;
   dp = opendir(mscclAlgoDir);
   if (dp == nullptr) {
-    // Try to find the algorithm directory under share folder based on librccl.so path
+    // Try to find the algorithm directory under share folder based on libnccl.so or librccl.so path
```
Comment (yzygitzh) [inline src/scheduler.cc:107]: Should be based on libmsccl-scheduler?

File: src/scheduler.cc [hunk 8]
```diff
@@ -65,6 +114,7 @@ __hidden ncclResult_t mscclSchedulerInit() {
   } else {
     fullDirPath = mscclAlgoDir;
   }
+  fprintf(stdout, "Using MSCCL Algo files from %s \n", fullDirPath);
```
Comment (yzygitzh) [inline src/scheduler.cc:117]: Format align with line 90

File: src/scheduler.cc [hunk 9]
```diff
@@ -35,27 +83,28 @@ __hidden ncclResult_t mscclSchedulerInit() {
...
-    mscclAlgoDirStr += (mscclUnitTestMode && mscclUnitTestMode()) ? mscclUnitTestAlgoDefaultDir : mscclAlgoDefaultDir;
+    mscclAlgoDirStr += (mscclUnitTestMode && mscclUnitTestMode()) ? mscclUnitTestAlgoDefaultDir : updateAlgoDirByVMSize(std::string(mscclAlgoDefaultDir));
     mscclAlgoDir = mscclAlgoDirStr.c_str();
     // Get share Directory Paths
     mscclAlgoShareDirStr = selfLibPath.substr(0, selfLibPath.find("lib"));
-    mscclAlgoShareDirStr += (mscclUnitTestMode && mscclUnitTestMode()) ? mscclUnitTestAlgoShareDirPath : mscclAlgoShareDirPath;
+    mscclAlgoShareDirStr += (mscclUnitTestMode && mscclUnitTestMode()) ? mscclUnitTestAlgoShareDirPath : updateAlgoDirByVMSize(std::string(mscclAlgoShareDirPath));
     mscclAlgoShareDir = mscclAlgoShareDirStr.c_str();
   }
+  fprintf(stdout, "MSCCL: External Scheduler will use %s as algorithm directory and %s as share algorithm directory\n", mscclAlgoDir, mscclAlgoShareDir);
```
Comment (yzygitzh) [inline src/scheduler.cc:102]: Format align with line 90

File: README.md (hunk 1 above)
Comment (Andyli1007) [reply README.md]: fixed.

File: src/scheduler.cc (hunk 2 above)
Comment (Andyli1007) [reply src/scheduler.cc:33]: added.

File: src/scheduler.cc (hunk 4 above)
Comment (Andyli1007) [reply src/scheduler.cc:96]: will keep the symbol named mscclSchedulerInit used for search for the scheduler.so, but already keep align the so file find path pattern.

File: src/scheduler.cc (hunk 5 above)
Comment (Andyli1007) [reply src/scheduler.cc]: changed accordingly.

File: src/scheduler.cc (hunk 6 above)
Comment (Andyli1007) [reply src/scheduler.cc]: changed accordingly.

File: src/scheduler.cc (hunk 8 above)
Comment (Andyli1007) [reply src/scheduler.cc]: changed.

File: src/scheduler.cc (hunk 9 above)
Comment (Andyli1007) [reply src/scheduler.cc]: changed.

File: pkg/debian/rules [hunk 10]
```diff
@@ -2,7 +2,7 @@
 
 %:
 	dh $@ --parallel
-
+
```
Comment (yzygitzh) [inline pkg/debian/rules:5]: Strange space

File: src/scheduler.cc [hunk 11]
```diff
@@ -95,26 +98,27 @@ __hidden ncclResult_t mscclSchedulerInit() {
...
+  fprintf(stdout, "%s: External Scheduler will use %s as algorithm directory and %s as share algorithm directory\n", MSCCL_SCHEDULER_NAME, mscclAlgoDir, mscclAlgoShareDir);
   struct dirent *entry = nullptr;
   DIR *dp = nullptr;
   dp = opendir(mscclAlgoDir);
   if (dp == nullptr) {
-    // Try to find the algorithm directory under share folder based on libnccl.so or librccl.so path
+    // Try to find the algorithm directory under share folder based on libmsccl-scheduler.so path
     dp = opendir(mscclAlgoShareDir);
     if (dp == nullptr) {
+      perror("opendir failed");
```
Comment (yzygitzh) [inline src/scheduler.cc:113]: why we need another error print?

File: pkg/debian/rules (hunk 10 above)
Comment (Andyli1007) [reply pkg/debian/rules]: fixed.

File: src/scheduler.cc (hunk 11 above)
Comment (Andyli1007) [reply src/scheduler.cc]: it's debugging purpose. already removed now.

File: src/scheduler.cc [hunk 12]
```diff
@@ -59,7 +63,7 @@ static std::string updateAlgoDirByVMSize(std::string algoDir){
         curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT, 1L);
         res = curl_easy_perform(curl);
         if(res != CURLE_OK) {
-            fprintf(stderr, "%s: Get Azure VM Size failed: %s", MSCCL_SCHEDULER_NAME, curl_easy_strerror(res));
+            fprintf(stdout, "%s: %s Get Azure VM Size failed: %s \n", MSCCL_SCHEDULER_NAME, LOG_WARN, curl_easy_strerror(res));
```
Comment (yzygitzh) [inline src/scheduler.cc:66]: Remove spaces before \n

**Review by yzygitzh — APPROVED** [review]

File: src/scheduler.cc (hunk 12 above)
Comment (Andyli1007) [reply src/scheduler.cc]: done.

**Pull request merged**